core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java

```diff
@@ -0,0 +1,404 @@
+  public SimpleInodeLockList(InodeLockManager inodeLockManager) {
+    mInodeLockManager = inodeLockManager;
+    mEntries = new ArrayList<>(INITIAL_CAPACITY);
+    mFirstWriteLockIndex = 0;
+  }
+
+  @Override
+  public void lockInode(Inode inode, LockMode mode) {
+    mode = nextLockMode(mode);
+    if (!mEntries.isEmpty()) {
+      Preconditions.checkState(!endsInInode(),
+          "Cannot lock inode %s for lock list %s because the lock list already ends in an inode",
+          inode.getId(), this);
+      String lastEdgeName = ((EdgeEntry) lastEntry()).getEdge().getName();
+      Preconditions.checkState(inode.getName().equals(lastEdgeName),
+          "Expected to lock inode %s but locked inode %s", lastEdgeName, inode.getName());
+    }
+    addEntry(createInodeEntry(inode, mode));
+  }
+
+  @Override
+  public void lockEdge(Inode lastInode, String childName, LockMode mode) {
+    mode = nextLockMode(mode);
+    long edgeParentId = lastInode.getId();
+    Edge edge = new Edge(lastInode.getId(), childName);
+    if (!mEntries.isEmpty()) {
+      Preconditions.checkState(endsInInode(),
+          "Cannot lock edge %s when lock list %s already ends in an edge", edge, this);
+      Preconditions.checkState(lastInode().getId() == edgeParentId,
+          "Cannot lock edge %s when the last inode id in %s is %s", edge, this, lastInode.getId());
+    }
+    addEntry(createEdgeEntry(edge, mode));
+  }
+
+  @Override
+  public void lockRootEdge(LockMode mode) {
+    Preconditions.checkState(mEntries.isEmpty(),
+        "Cannot lock root edge when lock list %s is nonempty", this);
+    addEntry(createEdgeEntry(ROOT_EDGE, mode));
+  }
+
+  @Override
+  public void pushWriteLockedEdge(Inode inode, String childName) {
+    Edge edge = new Edge(inode.getId(), childName);
+    Preconditions.checkState(!endsInInode(),
+        "Cannot push edge write lock to edge %s; lock list %s ends in an inode", edge, this);
+    Preconditions.checkState(endsInWriteLock(),
+        "Cannot push write lock; lock list %s ends in a read lock");
+
+    if (endsInMultipleWriteLocks()) {
+      // If the lock before the edge lock is already WRITE, we can just acquire more WRITE locks.
+      lockInode(inode, LockMode.WRITE);
+      lockEdge(inode, childName, LockMode.WRITE);
+    } else {
+      EdgeEntry lastEdgeEntry = createEdgeEntry(lastEdge(), LockMode.READ);
+      InodeEntry inodeEntry = createInodeEntry(inode, LockMode.READ);
+      EdgeEntry nextEdgeEntry = createEdgeEntry(edge, LockMode.WRITE);
+      removeLastEntry(); // Remove edge write lock
+      addEntry(lastEdgeEntry);
+      addEntry(inodeEntry);
+      addEntry(nextEdgeEntry);
+    }
+  }
+
+  @Override
+  public void unlockLastInode() {
+    Preconditions.checkState(endsInInode(),
+        "Cannot unlock last inode when the lock list %s ends in an edge", this);
+    Preconditions.checkState(!mEntries.isEmpty(),
+        "Cannot unlock last inode when the lock list is empty");
+    removeLastEntry();
```

Comment:
Will it be better to swap these two `checkState`? 
since `endsInInode==false` when `mEntries.isEmpty` or `endsInEdge`.